src/jit/lsraxarch.cpp

```diff
@@ -573,210 +698,56 @@ void LinearScan::BuildNode(GenTree* tree)
-            {
-                // These tree nodes will have their op1 marked as isDelayFree=true.
-                // Hence these tree nodes should have a Def position so that op1's reg
-                // gets freed at DefLoc+1.
-                if (tree->TypeGet() == TYP_VOID)
-                {
-                    // Right now a GT_XADD node could be morphed into a
-                    // GT_LOCKADD of TYP_VOID. See gtExtractSideEffList().
-                    // Note that it is advantageous to use GT_LOCKADD
-                    // instead of of GT_XADD as the former uses lock.add,
-                    // which allows its second operand to be a contained
-                    // immediate wheres xadd instruction requires its
-                    // second operand to be in a register.
-                    assert(info->dstCount == 0);
-
-                    // Give it an artificial type and mark it as an unused value.
-                    // This results in a Def position created but not considered consumed by its parent node.
-                    tree->gtType        = TYP_INT;
-                    info->dstCount      = 1;
-                    info->isLocalDefUse = true;
-                    tree->SetUnusedValue();
-                }
-                else
-                {
-                    assert(info->dstCount != 0);
-                }
-
-                delayUseSrc = op1;
-            }
-            else if ((info->dstCount != 0) && (op2 != nullptr) &&
-                     (!tree->OperIsCommutative() || (op2->isContained() && !op2->IsCnsIntOrI())))
+            srcCount = BuildBinaryUses(tree->AsOp());
+            if (internalDef != nullptr)
             {
-                delayUseSrc = op2;
-            }
-            if ((delayUseSrc != nullptr) && CheckAndSetDelayFree(delayUseSrc))
-            {
-                info->hasDelayFreeSrc = true;
+                buildInternalRegisterUses();
             }
+            BuildDef(tree);
         }
-    }
+        break;
 
-    BuildCheckByteable(tree);
+    } // end switch (tree->OperGet())
 
-    // We need to be sure that we've set info->srcCount and info->dstCount appropriately
-    assert((info->dstCount < 2) || (tree->IsMultiRegCall() && info->dstCount == MAX_RET_REG_COUNT));
-    assert(info->isLocalDefUse == (tree->IsValue() && tree->IsUnusedValue()));
-    assert(!tree->IsUnusedValue() || (info->dstCount != 0));
-    assert(info->dstCount == tree->GetRegisterDstCount());
+    // We need to be sure that we've set srcCount and dstCount appropriately
+    assert((dstCount < 2) || (tree->IsMultiRegCall() && dstCount == MAX_RET_REG_COUNT));
+    assert(isLocalDefUse == (tree->IsValue() && tree->IsUnusedValue()));
+    assert(!tree->IsUnusedValue() || (dstCount != 0));
+    assert(dstCount == tree->GetRegisterDstCount());
+    INDEBUG(dumpNodeInfo(tree, dstCandidates, srcCount, dstCount));
+    return srcCount;
```

Comment:
It looks like the return value of this function is used only in an assert. I hope that assert is worth all the added complication :)

Reply:
Yes - it's something I think should probably be removed, but I found it useful for debugging and comparison as I was making the transition from `TreeNodeInfo`